devito/dle/backends/advanced.py

```diff
@@ -0,0 +1,434 @@
+        arguments = [BlockingArg(v, k, blockshape[k]) for k, v in blocked.items()]
+
+        return {'nodes': processed, 'arguments': arguments, 'flags': 'blocking'}
+
+    @dle_pass
+    def _simdize(self, state, **kwargs):
+        """
+        Add compiler-specific or, if not available, OpenMP pragmas to the
+        Iteration/Expression tree to emit SIMD-friendly code.
+        """
+        ignore_deps = as_tuple(self._compiler_decoration('ignore-deps'))
+
+        def decorate(nodes):
+            processed = []
+            for node in nodes:
+                mapper = {}
+                for tree in retrieve_iteration_tree(node):
+                    vector_iterations = [i for i in tree if i.is_Vectorizable]
+                    for i in vector_iterations:
+                        handle = FindSymbols('symbolics').visit(i)
+                        try:
+                            aligned = [j for j in handle
+                                       if j.shape[-1] % get_simd_items(j.dtype) == 0]
+                        except KeyError:
+                            aligned = []
+                        if aligned:
+                            simd = omplang['simd-for-aligned']
+                            simd = simd(','.join([j.name for j in aligned]),
+                                        simdinfo[get_simd_flag()])
+                        else:
+                            simd = omplang['simd-for']
+                        mapper[i] = List(ignore_deps + as_tuple(simd), i)
+                processed.append(Transformer(mapper).visit(node))
+            return processed
+
+        return {'nodes': decorate(state.nodes),
+                'elemental_functions': decorate(state.elemental_functions)}
+
+    @dle_pass
+    def _ompize(self, state, **kwargs):
+        """
+        Add OpenMP pragmas to the Iteration/Expression tree to emit parallel code
+        """
+
+        processed = []
+        for node in state.nodes:
+
+            # Reset denormals flag each time a parallel region is entered
+            denormals = FindNodes(Denormals).visit(state.nodes)
+            mapper = {i: List(c.Comment('DLE: moved denormals flag')) for i in denormals}
+
+            # Handle parallelizable loops
+            for tree in retrieve_iteration_tree(node):
+                # Determine the number of consecutive parallelizable Iterations
+                key = lambda i: i.is_Parallel and not i.is_Vectorizable
+                candidates = filter_iterations(tree, key=key, stop='consecutive')
+                if not candidates:
+                    continue
+
+                # Heuristic: if at least two parallel loops are available and the
+                # physical core count is greater than self.thresholds['collapse'],
+                # then omp-collapse the loops
+                nparallel = len(candidates)
+                if psutil.cpu_count(logical=False) < self.thresholds['collapse'] or\
```

Comment:
yes, I suggest we have a key-value pair in ``parameters`` such as ``(dle: {...}``, where the ``{...}`` is a dictionary with parameters concerning the DLE. Again, this PR just moves code around, so taking note for subsequent improvement